packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2127,6 +2127,11 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       return;
     }
 
+    if (value.selection.isCollapsed && value.selection.affinity != _value.selection.affinity) {
```

Comment:
Do you still want to revert the affinity change if the user actually typed something?

Reply:
As you point out that the input server never know the actually affinity and can not change the affinity by key board.
Currently we can only change the affinity by tap the text string and change the selection.
Additional, the affinity is only valid when line breaks cause by render wrapping, not explicit newline characters.

Reply:
Ah sorry I misspoke, the engine doesn't care but the framework does (for rendering the caret at the right place).

As far as I can tell on iOS/Android when the user types using a virtual/hardware keyboard the affinity is always set to `TextAffinity.downstream`. `TextAffinity.upstream` is only used when the selection is change by certain gestures when the caret position is located at a soft-wrap line boundary, or bidi boundary. For example when the text field needs to show 
```
aa|
bb
```
instead of 
```
aa
|bb
```

Reply:
Absolutely correct, so we can override the affinity from engine, right?
Otherwise, the only affinity change will trigger selection change and dismiss the handler by mistake.

Reply:
Say the line can only accommodate 2 characters, and the text editing state was initially 
`(text: 'a', textSelection.collapsed(1, affinity: upstream))`, when the user types another `a` the affinity should be changed to downstream?

Reply:
1,Type the second 'a', the engine will send a wrong affinity, but it does not affect the caret render.
`TextPainter._computeCaretMetrics` will find the real position.
```dart
I/flutter ( 7996): updateEditingValue old: TextEditingValue(text: ┤a├, selection: TextSelection.collapsed(offset: 1, affinity: TextAffinity.upstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I/flutter ( 7996): updateEditingValue new form engine: TextEditingValue(text: ┤aa├, selection: TextSelection.collapsed(offset: 2, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
```


2,Type the 3rd 'a' 
```dart
I/flutter ( 7996): updateEditingValue old: TextEditingValue(text: ┤aa├, selection: TextSelection.collapsed(offset: 2, affinity: TextAffinity.upstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I/flutter ( 7996): updateEditingValue new form engine: TextEditingValue(text: ┤aaa├, selection: TextSelection.collapsed(offset: 3, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
```

Reply:
> the engine will send a wrong affinity

`downstream` is the correct text affinity here when a second `a` is entered

Reply:
> `downstream` is the correct text affinity here when a second `a` is entered

Ah, Yes, currently, the cursor at the end of the line and the text did not render wrapping, so the affinity is invalid or meaningless.

In summary, all from engine's affinity is meaningless and we can override it, right?

Reply:
If the engine change actually contains text change or selection change then it should typically be downstream.

Reply:
Could you explain why? The documentation says that affinity only makes sense when there is ambiguity in the visual location, and is used to paint the caret. Am I miss something?

Reply:
When the user types something, or changes the selection using a keyboard, the caret typically shouldn't be painted using the `upstream` affinity. In other words it should be
```
aa
|bb
```
instead of
```
aa|
bb
```

Reply:
I almost understand what you mean.
Another case, if the `aa` is composed, and I wanna to type `aaa` with composed, it should be painted using upstream affinity?

Reply:
That I'm not sure. On Android it's behaving like it's upstream. But I vaguely remember there's no cursor affinity on Android.

Reply:
Done.